lib/multipass_ex/crypto.ex

```diff
@@ -0,0 +1,129 @@
+          {:ok, String.t()}
+          | {:error, EncodingError.t()}
+  def encrypt(data, site_key, api_key) when is_map(data) do
+    key = generate_key(site_key, api_key)
+
+    case prepare_data_for_encryption(data) do
+      {:ok, prepared_data} ->
+        encrypted_data =
+          @algorithm
+          |> :crypto.block_encrypt(key, @iv, prepared_data)
+          |> Coding.encode()
+
+        {:ok, encrypted_data}
+
+      {:error, error_struct} ->
+        {:error, error_struct}
+    end
+  end
+
+  @spec decrypt(String.t(), binary(), binary()) ::
+          {:ok, map()} | {:error, CryptoError.t() | DecodingError.t() | JSONDecodingError.t()}
+  def decrypt(data, site_key, api_key) do
+    case Coding.decode(data) do
+      {:ok, decoded_data} ->
+        decrypt_or_error(decoded_data, site_key, api_key)
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @spec decrypt_or_error(binary(), binary(), binary()) ::
+          {:ok, map()} | {:error, CryptoError.t() | DecodingError.t()}
+  defp decrypt_or_error(data, site_key, api_key) do
+    key = generate_key(site_key, api_key)
+
+    @algorithm
+    |> :crypto.block_decrypt(key, @iv, data)
+    |> process_decrypted_data()
+    |> case do
+      {:ok, data} -> {:ok, data}
+      {:error, error_struct} -> {:error, error_struct}
+    end
+  rescue
+    ArgumentError ->
+      {:error, %CryptoError{message: "decryption failed", reason: :argument_error}}
+  end
+
+  @spec process_decrypted_data(binary()) :: {:ok, map()} | {:error, JSONDecodingError.t()}
+  defp process_decrypted_data(data) do
+    data
+    |> unpad()
+    |> Jason.decode()
+    |> case do
+      {:ok, json} ->
+        {:ok, json}
+
+      {:error, _exception} ->
+        {:error,
+         %JSONDecodingError{
+           message: "json decoding failed",
+           reason: :json_decoding_error,
+           data: data
+         }}
+    end
+  end
+
+  @spec prepare_data_for_encryption(map()) :: {:ok, binary} | {:error, EncodingError.t()}
+  defp prepare_data_for_encryption(data) do
+    prepared_data =
+      data
+      |> Jason.encode!()
+      |> pad()
+
+    {:ok, prepared_data}
+  rescue
+    Jason.EncodeError ->
+      {:error,
+       %EncodingError{
+         message: "json encoding failed",
+         reason: :json_encoding_error
+       }}
+  end
+
+  @spec generate_key(String.t(), String.t()) :: binary()
+  defp generate_key(password, salt) do
+    :crypto.hash(:sha, salt <> password) |> truncate_secret()
+  end
+
+  # typespec: capture a binary followed by a bitstring and return 128 bits (@key_byte_length)
+  # credo:disable-for-next-line
+  @spec truncate_secret(<<_::8, _::_*1>>) :: <<_::128>>
+  defp truncate_secret(<<key::binary-size(@key_byte_length), _rest::bitstring>>) do
+    key
+  end
+
+  @spec pad(String.t()) :: binary()
+  defp pad(data) when is_binary(data) do
+    to_add = @block_size - rem(byte_size(data), @block_size)
+    data <> :binary.copy(<<to_add>>, to_add)
+  end
```

Comment:
I'm not quite sure what's happening here... It looks like if you're `12` bytes short, it appends the binary "12", 12 times? What happens if you're short by over 1 block size?

I'm not super familiar with padding and unpadding, so these might be dumb questions. 😅

Reply:
You are correct with the padding, also you won't be over 1 block size short as you would fall into the previous block size and pad their. Crude example below, x = data, 0 = padding.

```
|----|----|----|
 xxx0
 xxxx xx00
 xxxx xxxx x000
```

Reply:
now the real question is, what happens if the data size is an integer multiple of the block size?

Reply:
Wouldn't you add a full block of padding?

Reply:
yeah, definitely need to add a whole block. looks like ivy implemented pkcs#7 incorrectly

Reply:
err, sorry, i thought ivy had implemented it incorrectly.

Reply:
Is this code actually appending x bytes of `0`'s or x bytes of the binary version of whatever x is?

Reply:
It will be adding 16 bytes of the value 16.

```elixir
iex(4)> "frank" <> :binary.copy(<<16>>, 16)
<<102, 114, 97, 110, 107, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16,
  16, 16, 16>>
```

Reply:
Ok, I just wanted to make sure. Would padding 16 0's make more sense?

Reply:
https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS#5_and_PKCS#7

Reply:
Thanks for the link. TIL